emloop/api.py

```diff
@@ -0,0 +1,234 @@
+import os
+import errno
+import logging
+import os.path as path
+from datetime import datetime
+import copy
+from collections import namedtuple
+from typing import Optional, Iterable, List
+
+from .datasets import AbstractDataset
+from .models import AbstractModel
+from .hooks import AbstractHook
+from .hooks.training_trace import TrainingTrace
+from .cli.util import fallback
+from .constants import EL_LOG_FILE
+from .constants import EL_LOG_FILE, EL_HOOKS_MODULE, EL_CONFIG_FILE, EL_LOG_DATE_FORMAT, EL_LOG_FORMAT
+from .utils.reflection import get_class_module, parse_fully_qualified_name, create_object
+from .utils.yaml import yaml_to_str, yaml_to_file, load_yaml
+from .utils import get_random_name
+from .main_loop import MainLoop
+
+
+def create_output_dir(config: dict, output_root: str, default_model_name: str='Unnamed') -> str:
+    """
+    Create output_dir under the given ``output_root`` and
+        - dump the given config to YAML file under this dir
+        - register a file logger logging to a file under this dir
+
+    :param config: config to be dumped
+    :param output_root: dir wherein output_dir shall be created
+    :param default_model_name: name to be used when `model.name` is not found in the config
+    :return: path to the created output_dir
+    """
+    logging.info('Creating output dir')
+
+    model_name = default_model_name
+    if 'name' not in config['model']:
+        logging.warning('\tmodel.name not found in config, defaulting to: %s', model_name)
+    else:
+        model_name = config['model']['name']
+
+    if not os.path.exists(output_root):
+        logging.info('\tOutput root folder "%s" does not exist and will be created', output_root)
+        os.makedirs(output_root)
+
+    output_dir_format = os.environ.get('EMLOOP_OUTPUT_DIR_FORMAT', '{timestamp}_{model_name}_{random_name}')
+    timestamp = datetime.now().strftime('%Y-%m-%d-%H-%M-%S')
+    random_name = get_random_name()
+    output_dir_name = output_dir_format.format(timestamp=timestamp, model_name=model_name, random_name=random_name)
+
+    suffix = 0
+    while True:
+        if suffix > 0:
+            output_dir_name = output_dir_name + f'_{suffix}'
+
+        output_dir = os.path.join(output_root, output_dir_name)
+        try:
+            os.mkdir(output_dir)
+            break
+        except FileExistsError:
+            suffix += 1
+
+    logging.info(f'Created output directory with name {output_dir}')
+    logging.info('\tOutput dir: %s', output_dir)
+    yaml_to_file(data=config, output_dir=output_dir, name=EL_CONFIG_FILE)
+
+    # create file logger
```

Comment:
This would mean adding new function to the API which saves the config. It would allow the user to create an output directory without a `config.yaml` file. But I cannot imagine a situation where it would be useful.  So the user of the API would always call `create_output_dir` then `save_config`, unless he forgets and then the information about the training would be lost.